models/user.go

```diff
@@ -1 +1,337 @@
+		return InvalidUserError("auth_data_pwd", u.ID)
+	}
+
+	if len(u.Password) > USER_PASSWORD_MAX_LENGTH {
+		return InvalidUserError("password_limit", u.ID)
+	}
+
+	if !IsValidLocale(u.Locale) {
+		return InvalidUserError("locale", u.ID)
+	}
+
+	if len(u.Timezone) > 0 {
+		if tzJSON, err := json.Marshal(u.Timezone); err != nil {
+			return NewAppError("User.IsValid", "model.user.is_valid.marshal.app_error", nil, err.Error(), http.StatusInternalServerError)
+		} else if utf8.RuneCount(tzJSON) > USER_TIMEZONE_MAX_RUNES {
+			return InvalidUserError("timezone_limit", u.ID)
+		}
+	}
+
+	if len(u.ContactNumber) > USER_CONTACT_NUMBER_MAX_LENGTH {
+		return InvalidUserError("contact_number_limit", u.ID)
+	}
+
+	return nil
+}
+
+func InvalidUserError(fieldName string, userId string) *AppError {
+	id := fmt.Sprintf("model.user.is_valid.%s.app_error", fieldName)
+	details := ""
+	if userId != "" {
+		details = "user_id=" + userId
+	}
+	return NewAppError("User.IsValid", id, nil, details, http.StatusBadRequest)
+}
+
+func NormalizeUsername(username string) string {
+	return strings.ToLower(username)
+}
+
+func NormalizeEmail(email string) string {
+	return strings.ToLower(email)
+}
+
+// PreSave will set the Id and Username if missing.  It will also fill
+// in the CreatedAt, UpdatedAt times.  It will also hash the password.  It should
+// be run before saving the user to the db.
+func (u *User) PreSave() {
+	if u.ID == "" {
+		u.ID = NewId()
+	}
+
+	if u.Username == "" {
+		u.Username = NewId()
+	}
+
+	if u.AuthData != nil && *u.AuthData == "" {
+		u.AuthData = nil
+	}
+
+	u.Username = SanitizeUnicode(u.Username)
+	u.FirstName = SanitizeUnicode(u.FirstName)
+	u.LastName = SanitizeUnicode(u.LastName)
+
+	u.Username = NormalizeUsername(u.Username)
+	u.Email = NormalizeEmail(u.Email)
+
+	u.CreatedAt = GetMillis()
+	u.UpdatedAt = u.CreatedAt
+
+	if u.Locale == "" {
+		u.Locale = DEFAULT_LOCALE
+	}
+
+	if u.Timezone == nil {
+		u.Timezone = timezones.DefaultUserTimezone()
+	}
+
+	if u.Password != "" {
+		u.Password = HashPassword(u.Password)
+	}
+}
+
+// PreUpdate should be run before updating the user in the db.
+func (u *User) PreUpdate() {
+	u.Username = SanitizeUnicode(u.Username)
+	u.FirstName = SanitizeUnicode(u.FirstName)
+	u.LastName = SanitizeUnicode(u.LastName)
+
+	u.Username = NormalizeUsername(u.Username)
+	u.Email = NormalizeEmail(u.Email)
+	u.UpdatedAt = GetMillis()
+
+	u.FirstName = SanitizeUnicode(u.FirstName)
+	u.LastName = SanitizeUnicode(u.LastName)
+
+	if u.AuthData != nil && *u.AuthData == "" {
+		u.AuthData = nil
+	}
+
+}
+
+func (u *User) Patch(patch *UserPatch) {
+	if patch.Username != nil {
+		u.Username = *patch.Username
+	}
+
+	if patch.FirstName != nil {
+		u.FirstName = *patch.FirstName
+	}
+
+	if patch.LastName != nil {
+		u.LastName = *patch.LastName
+	}
+
+	if patch.Email != nil {
+		u.Email = *patch.Email
+	}
+
+	if patch.Locale != nil {
+		u.Locale = *patch.Locale
+	}
+
+	if patch.Timezone != nil {
```

Comment:
Probably we'll ignore or remove this here too if we plan to get rid of Timezone management.